Introduce Compaction Groups

raphaelsc opened the pull request:
Compaction group can be defined as a set of files that can be compacted together. Today, all sstables belonging to a table in a given shard belong to the same group. So we can say there's one group per table per shard. As we want to eventually allow isolation of data that shouldn't be mixed, e.g. data from different vnodes, then we want to have more than one group per table per shard. That's why compaction groups is being introduced here.

Today, all memtables and sstables are stored in a single structure per table. After compaction groups, there will be memtables and sstables for each group in the table.

As we're taking an incremental approach, table still supports a single group. But work was done on preparing table for supporting multiple groups. Completing that work is actually the next step. Also, a procedure for deriving the group from token is introduced, but today it always return the single group owned by the table. Once multiple groups are supported, then that procedure should be implemented to map a token to a group.

No semantics was changed by this series.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1753/

avikivity commented on replica/database.hh:
```diff
@@ -555,7 +552,7 @@ private:
     do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::shared_sstable sstable,
         enable_backlog_tracker backlog_tracker);
-    void add_sstable(sstables::shared_sstable sstable);
+    void add_sstable(compaction_group& cg, sstables::shared_sstable sstable);
```
I expected to see table::add_sstable(sstable) that figures out which compaction_group to talk to, and compaction_group::add_sstable(sstable) that does the work. How does the caller of add_sstable(compaction_group, sstable) know to select the correct compaction_group? Or to split the sstable if needed?

raphaelsc replied:
table::add_sstable(compaction_group& cg, sstables::shared_sstable sstable) is a private function. I think you're referring to table::add_sstable_and_update_cache() which is in the public namespace. And yes, it should figure out which group a file belongs to. In this series, add_sstable_and_update_cache() is adding the sstable on behalf of the single group. With my next step, which is to add support to multiple groups, then a function will indeed be introduced to figure out the ownership of the sstable file.

avikivity commented on replica/table.cc:
```diff
@@ -405,8 +407,8 @@ table::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::s
-void table::add_sstable(sstables::shared_sstable sstable) {
-    _main_sstables = do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
+void table::add_sstable(compaction_group& cg, sstables::shared_sstable sstable) {
+    cg._main_sstables = do_add_sstable(cg._main_sstables, std::move(sstable), enable_backlog_tracker::yes);
```
This really should delegate to compaction_group so it is control of what is happening. Maybe compaction_group shouldn't let table reach into its internals (but vice versa is okay).

raphaelsc replied:
Makes sense. table::add_sstable(compaction_group, ...) should become compaction_group::add_sstable(...). Same logic applying elsewhere.

raphaelsc commented:
v2:
- table no longer have access to compaction_group private namespace
- fixed scylla-gdb memtables
- fast path in table::apply(const frozen_mutation& m, ...) if single group, to avoid unnecessary token calculation.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1897/

raphaelsc commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1897/

Failed due to a flakiness in test/raft/randomized_nemesis_test. Already reported to raft devs.

bhalevy commented on replica/database.hh:
```diff
@@ -318,6 +318,7 @@ struct table_stats;
 using column_family_stats = table_stats;
 
 class database_sstable_write_monitor;
+class compaction_group;
```
Shouldn't it be in the sstables namespace?
(Maybe we need a compaction namespace by the way, but that's orthogonal to this change)

bhalevy replied:
Actually, I see that `table_state` is already defined in namespace compaction.
It makes sense to do the same for `compaction_group`.

raphaelsc replied:
The idea is that table_state becomes compaction_group_state. compaction_group is a replica entity, rather than one which lives in compaction namespace. We did tons of effort in compaction layer to not reference table directly, but rather work with the table_state. compaction group stores memtable_list and sstable_set's, all of which belong to the replica namespace.

I was thinking yesterday about your question, and perhaps I should create replica/compaction_group.hh and replica/compaction_group.cc, which allows for better organization of table and compaction group code.

Sounds good to you?

bhalevy replied:
The organization makes sense.
Unfortunately the name doesn't.
I think this entity is more than about compaction so it's name is misleading.
For example this group of tokens will be a unit for replication purposes, and consequently, for repair.  Eventually, it will become a tablet.

raphaelsc replied:
"subtable (plural subtables) A table of data corresponding to part of a greater parent table." ? :-) 

table_state can become compaction_group (makes sense in the compaction layer), but you're right compaction_group is not a good name for that class, living in replica namespace, storing a subset of a table.

raphaelsc replied:
Compaction group could still be a good name if we think that compaction group refers to a LSM storage tree that is isolated and can be operated independently from others which belong to the same table. Then naturally compaction group interface will expose method for interacting with its memtable (which is in first level of the tree) and sstable set(s)  (which belong to higher levels).

bhalevy replied:
Maybe they can be already be called a tablet or simple_tablet. @avikivity what do you think?

bhalevy replied:
Or would it be appropriate only with tablet partitioning / replication mapping?

bhalevy replied:
Let's not hold our horses on naming.
I don't object to proceeding with `compaction_group`

avikivity replied:
Let's go with compaction_group. We can always rename it later.

bhalevy commented on replica/database.hh:
```diff
@@ -409,6 +410,7 @@ private:
     lw_shared_ptr<memtable_list> make_memory_only_memtable_list();
     lw_shared_ptr<memtable_list> make_memtable_list();
 
+    compaction_manager& _compaction_manager;
```
nit: why moving _compaction_manager here is needed?

raphaelsc replied:
That's because table::_compaction_manager should be set before group accesses it.

bhalevy replied:
ok

bhalevy commented on replica/database.hh:
```diff
@@ -1102,6 +1105,25 @@ public:
+
+// Compaction group is a set of SSTables which are eligible to be compacted together.
+// By this definition, we can say:
+//      - A group contains SSTables from the same table and shard
```
Aren't we currently serving each compaction_group by all shards?

bhalevy commented on replica/database.hh:
```diff
@@ -1102,6 +1105,25 @@ public:
+    compaction_group(table& t);
+
+    // Will stop ongoing compaction on behalf of this group, etc.
+    future<> stop();
```
noexcept please.
exception in the `stop()` path are evil.

raphaelsc replied:
stop() does flush a memtable, shouldn't we fail gracefully rather than aborting?

bhalevy replied:
Even if flush may fail gracefully, we should return that as an exceptional future.

bhalevy commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -444,12 +444,13 @@ static future<std::vector<unsigned long>> compact_sstables(test_env& env, sstrin
-                column_family_test(cf).add_sstable(sst);
+                candidates.push_back(sst);
             }
-            auto candidates = get_candidates_for_leveled_strategy(*cf);
```
let's break this into a patch of its own.
and justify it there.

raphaelsc replied:
That change was required as part of the change, as that unit test doesn't run in a thread, and get_candidates_for_leveled_strategy() would fail into is_running_in_a_thread assertion.

raphaelsc replied:
I can move this change into a separate patch that comes before the change in column_family_test.

bhalevy replied:
thanks

bhalevy commented on test/boost/sstable_test.hh:
```diff
@@ -32,18 +32,16 @@ class column_family_test {
-        auto builder = replica::table::sstable_list_builder(std::move(permit));
-        _cf->_main_sstables = builder.build_new_list(*_cf->_main_sstables, _cf->_compaction_strategy.make_sstable_set(_cf->schema()), new_sstables, sstables_to_remove).get0();
-        _cf->refresh_compound_sstable_set();
+        _cf->as_table_state().on_compaction_completion(sstables::compaction_completion_desc{ .old_sstables = sstables_to_remove, .new_sstables = new_sstables }, sstables::offstrategy::no).get();
```
This change seems intrusive enough to be separated into its own patch too.

raphaelsc replied:
All the changes in "test: compaction: Prepare tests for compaction group changes" were done as a consequence of these change in column_family_test. That's because those functions now have to run in a thread. But I can prepare the tests for the preparation of the tests :-)

bhalevy commented on test/lib/sstable_utils.hh:
```diff
@@ -213,6 +213,8 @@ public:
 
     void set_values(sstring first_key, sstring last_key, stats_metadata stats) {
+        _sst->_data_file_size = 1;
+        _sst->_bytes_on_disk = 1;
```
This too should be separated out and explained.
It seems to be valid even without compaction_groups.

raphaelsc replied:
Yes, they can go into a separate patch. thanks

bhalevy commented on replica/database.hh:
```diff
@@ -412,10 +412,10 @@ private:
     // Holds SSTables created by maintenance operations, which need reshaping before integration into the main set
     lw_shared_ptr<sstables::sstable_set> _maintenance_sstables;
+    // Holds single compaction group owned by table, meaning that all SSTables can be compacted with one another.
+    std::unique_ptr<compaction_group> _compaction_group;
```
I suggest placing the member at its final position already in the first patch that introduced it
to reduce the churn.

Right after `_compaction_strategy` seems even more appropriate.

raphaelsc replied:
I probably needed to put it after maintenance set because compaction_group ctor accessed table::_maintanence_sstables and found it uninitialized. But by the end of the series, _compaction_group will reach that location as those fields are moved into compaction group class.

bhalevy replied:
ok, maybe add a FIXME comment to set the stage for that

bhalevy commented on replica/database.hh:
```diff
@@ -1116,13 +1115,24 @@ public:
+    void add_sstable(sstables::shared_sstable sstable);
+
+    const lw_shared_ptr<sstables::sstable_set>& main_sstables() const;
+    void update_main_sstables(lw_shared_ptr<sstables::sstable_set> new_set);
```
nit: rename to `set_main_sstables`. `update` may imply the new_set is added into the existing set rather than replacing it.

raphaelsc replied:
Makes sense. thanks!

bhalevy commented on replica/table.cc:
```diff
@@ -405,9 +407,17 @@ table::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::s
+    _t.refresh_compound_sstable_set();
+}
+
+const lw_shared_ptr<sstables::sstable_set>& compaction_group::main_sstables() const {
```
nit: can be noexcept

raphaelsc replied:
will add

bhalevy commented on replica/table.cc:
```diff
@@ -405,9 +407,17 @@ table::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::s
-    refresh_compound_sstable_set();
+void compaction_group::add_sstable(sstables::shared_sstable sstable) {
+    _main_sstables = _t.do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
+    _t.refresh_compound_sstable_set();
```
This violates layering.
If that's (hopefully) temporary just for doing an incremental conversion of the code,
let's add a FIXME comment explaining and documenting it.

raphaelsc replied:
Yes, I really dislike this too. When writing this series, I tried to limit the amount of changes, and this was one of them. But I think I can move _t.refresh_compound_sstable_set() to be called by table itself, rather than relying on compaction group to do so.

bhalevy commented on replica/table.cc:
```diff
@@ -954,7 +971,7 @@ table::update_sstable_lists_on_off_strategy_completion(sstables::compaction_comp
 // Note: must run in a seastar thread
 void
-table::on_compaction_completion(sstables::compaction_completion_desc desc) {
+table::on_compaction_completion(compaction_group& cg, sstables::compaction_completion_desc desc) {
```
My gut feeling is that this method should move to compaction_group
and that should interact with the table layer to execute the table-wide code.

raphaelsc replied:
Shouldn't it be table coordinating the activity rather than the other way around? 

the update of sstable set happens through row_cache::invalidate(external_updater, ...) method, and we have table then refreshing its cache snapshot, updating its statistics, submitting files for deletion, etc.

Maybe sstable_list_updater (which implements the actual replacement), which is defined in table::on_compaction_completion() should be moved to compaction_group. Maybe it's not needed.

bhalevy replied:
I think that compaction_manager should deal with compaction_group_state that should interact with the compaction_group object, and if that needs interaction with the table, so be it.
But bypassing that and calling the table with a refwrence to the compaction group seems like violating layering.

bhalevy commented on replica/database.hh:
```diff
@@ -1129,10 +1127,15 @@ public:
 
+    const lw_shared_ptr<sstables::sstable_set>& maintenance_sstables() const;
+    void update_maintenance_sstables(lw_shared_ptr<sstables::sstable_set> new_set);
+
```
ditto, rename to `set_maintenance_sstables`

raphaelsc replied:
Will do.

bhalevy commented on replica/table.cc:
```diff
@@ -420,9 +420,17 @@ void compaction_group::update_main_sstables(lw_shared_ptr<sstables::sstable_set>
-    refresh_compound_sstable_set();
+void compaction_group::add_maintenance_sstable(sstables::shared_sstable sst) {
+    _maintenance_sstables = _t.do_add_sstable(_maintenance_sstables, std::move(sst), enable_backlog_tracker::no);
+    _t.refresh_compound_sstable_set();
```
ditto FIXME

bhalevy commented on replica/table.cc:
```diff
@@ -420,9 +420,17 @@ void compaction_group::update_main_sstables(lw_shared_ptr<sstables::sstable_set>
+    _t.refresh_compound_sstable_set();
+}
+
+const lw_shared_ptr<sstables::sstable_set>& compaction_group::maintenance_sstables() const {
```
ditto noexcept :)

bhalevy commented on replica/compaction_group.hh:
```diff
@@ -0,0 +1,75 @@
+// By this definition, we can say:
+//      - A group contains SSTables from the same table and shard
+//      - Each group can be thought of an isolated LSM tree.
+//      - Time Window strategy has one group per window
```
Doesn't that conflict with major compaction?

raphaelsc replied:
No, because there's no token overlapping among groups, so we can serialize major across groups

bhalevy replied:
> there's no token overlapping among groups

How come? Don't all groups in TWCS cover the whole token range and we just segregate the rows across groups based on the time window?

raphaelsc replied:
Ah you're talking about TWCS. That was just a conceptual statement. Will remove it to not confuse anybody else.

bhalevy replied:
Looks like the confusing comment is still here.
I don't know what you mean by "Time Window strategy" here, other than TWCS.

bhalevy commented on replica/compaction_group.hh:
```diff
@@ -0,0 +1,75 @@
+    compaction_group(table& t);
+
+    // Will stop ongoing compaction on behalf of this group, etc.
+    future<> stop();
```
noexcept please

bhalevy replied:
So we can/should make flush noexcept too, for exactly that reason.
It should be a very rare reason to allow an async function to fail in 2 ways - throwing an exception, and returning a failed future, because this can double the error handling efforts (in the outermost calls.  if a function throws in a continuation we convert that into an exceptional future).

The most common technique to use is:
```c++
future<> compaction_group::stop() {
  try {
    return _memtables->flush().finally([this] {
        return _t._compaction_manager.remove(as_table_state());
    });
  } catch (...) {
    return current_exception_as_future<>();
  }
}
```

This will not terminate.

Another way to do it is:
`futurize_invoke([this] { return _memtables->flush(); }).finally( ... )`

but it's cumbersome and ugly in this case.

So I'd vote for making flush noexcept to start with, possibly catching any exceptions there as described above.

raphaelsc replied:
That's orthogonal to this series, I think. I think I can start with your suggestion to convert any thrown exception into exception future and add a FIXME for the subsequent work.

raphaelsc replied:
That = make flush noexcept.

raphaelsc commented:
v3:
- rebased
- split preparatory changes for unit tests
- moved compaction_group into its dedicated header
- added noexcept qualified to methods retrieving sstable set in compaction_group
- moved functions called on compaction completion to group
- moved min_memtable_timestamp to group

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2089/

raphaelsc commented on replica/database.hh:
```diff
@@ -729,7 +726,9 @@ public:
     using const_row_ptr = std::unique_ptr<const row>;
-    memtable& active_memtable() { return _memtables->active_memtable(); }
+    // FIXME: for supporting multiple compaction groups, this interface will need to return
+    // as many active sstables are there are groups.
```
> nit: s/are there/as there/

thanks

raphaelsc commented:
v4:
- fixed grammar mistake in comment
- made compaction_group::stop() noexcept
- added patch to restore indentation in table::stop()

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2092/

bhalevy commented on test/boost/sstable_test.hh:
```diff
@@ -32,18 +32,16 @@ class column_family_test {
-        _cf->_main_sstables->insert(std::move(sstable));
-        _cf->refresh_compound_sstable_set();
+        auto new_sstables = { sstable };
+        _cf->as_table_state().on_compaction_completion(sstables::compaction_completion_desc{ .new_sstables = new_sstables }, sstables::offstrategy::no).get();
```
wouldn't it be simpler to return a `future<>` here and let the caller `.get()` it if it's in a thread?

bhalevy commented on replica/database.hh:
```diff
@@ -770,10 +774,15 @@ public:
+            return;
+        }
+        dht::token t = dht::get_token(*m_schema, m.key());
+        do_apply(compaction_group_for_token(t), std::move(h), m, m_schema);
```
We can have a variant of `compaction_group_for_token` taking references to the schema and key that will hide the complexity of `single_compaction_group_if_available`, doing essentially what this function is doing, just reusable in further use cases that may need it.

raphaelsc replied:
Makes sense. Can be named compaction_group_for_key().

bhalevy commented on replica/database.hh:
```diff
@@ -402,19 +403,17 @@ private:
-    lw_shared_ptr<sstables::sstable_set> _maintenance_sstables;
-    // Compound set which manages all the SSTable sets (e.g. main, etc) and allow their operations to be combined
+    // TODO: Still holds a single compaction group, meaning all sstables are eligible to be compacted with one another. Soon, a table
+    //  will be able to hold more than one group.
```
How about keeping a map of compaction_group per token range
and populate it with a single entry for the full range?

This will show the basic requirements of supporting multiple groups upfront
and will reduce the churn of supporting multiple groups later on.

avikivity replied:
I think it's okay to start with unique_ptr. Since there's just one now all the code to locate the correct group will be buggy.

avikivity commented on replica/table.cc:
```diff
@@ -1216,6 +1217,21 @@ table::make_memtable_list() {
 
+compaction_group::compaction_group(table& t)
+    : _t(t)
+    , _table_state(std::make_unique<table_state>(t))
```
Where is compaction_group::table_state defined?

raphaelsc replied:
Justed moved table::table_state to compaction group, so it's still defined in replica/table.cc while being renamed. One of the reasons is that it makes sense, and another reason is to avoid including database.hh as I cannot forward declare a nested class type.

avikivity commented on replica/table.cc:
```diff
@@ -1216,6 +1217,21 @@ table::make_memtable_list() {
+
+future<> compaction_group::stop() noexcept {
+    try {
+        return _t._compaction_manager.remove(as_table_state());
```
Better to change compaction_manager::remove to be noexcept.

raphaelsc replied:
Will do.

avikivity commented on replica/table.cc:
```diff
@@ -2379,7 +2394,7 @@ table::as_mutation_source_excluding(std::vector<sstables::shared_sstable>& ssts)
 }
 
-class table::table_state : public compaction::table_state {
+class compaction_group::table_state : public compaction::table_state {
```
How did it compile if it's defined after it's used?

raphaelsc replied:
Both GCC and clang allow me to instantiate the *ctor* of class that wasn't defined yet. Had an member been accessed instead, then the compiler would yell at me.

This is what we have:
https://godbolt.org/z/4eTs8n73P

raphaelsc replied:
With make_unique<T>(args...), we satisfy the compiler, then the class is found defined by the time the template is evaluated.

raphaelsc replied:
Interesting reading about this topic: https://howardhinnant.github.io/incomplete.html

avikivity replied:
ok

avikivity commented:
Looks good.

raphaelsc commented:
v5:
- made compaction_manager::remove() noexcept
- introduced table::compaction_group_for_key() to avoid the duplication of fast path (when table owns a single group)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2107/

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -789,7 +789,9 @@ future<> compaction_manager::stop_tasks(std::vector<shared_ptr<task>> tasks, sst
 
-future<> compaction_manager::stop_ongoing_compactions(sstring reason, compaction::table_state* t, std::optional<sstables::compaction_type> type_opt) {
+future<> compaction_manager::stop_ongoing_compactions(sstring reason, compaction::table_state* t, std::optional<sstables::compaction_type> type_opt) noexcept {
+  // FIXME: restore indentation
```
let's either fix the indentation in this series, or drop the comment, if we can live with half-indent here :)

raphaelsc replied:
Will add a patch to restore indentation.

bhalevy commented on replica/table.cc:
```diff
@@ -803,17 +862,14 @@ table::stop() {
+                        _sstables = make_compound_sstable_set();
+                    })).then([this] {
+                        _compaction_strategy.get_backlog_tracker().disable();
+                        _cache.refresh_snapshot();
```
nit: I think that if make this a coroutine as a preliminary change, it will be more obvious what exactly changed by reducing the churn and leaving only the parts still relevant to table and what moves to compaction_group.

raphaelsc replied:
In the first patch which touches table::stop(), I used that technique of breaking indentation to make it easy to verify what changed, then restored indentation in a later patch. So I assume that making it a couroutine is not needed for this specific reason. But coroutine has its benefits so I will convert table::stop() into coroutine in preliminary patch.

bhalevy reviewed (approved):
looks mostly good to me.
I had only a few minor comments

avikivity reviewed (approved):
Please add before/after perf_simple_query. I don't expect surprises, but this touches the hot path.

raphaelsc commented:
v6:
- Added patch to restore indentation in compaction_manager::stop_ongoing_compactions()
- Converted table::stop() into coroutine in a preliminary patch
- Improved documentation of compaction_group

raphaelsc commented:
> Please add before/after perf_simple_query. I don't expect surprises, but this touches the hot path.

before:

median 951913.28 tps ( 72.2 allocs/op,  14.3 tasks/op,   58052 insns/op,        0 errors)
median 958138.97 tps ( 72.2 allocs/op,  14.3 tasks/op,   57940 insns/op,        0 errors)
median 933679.25 tps ( 72.3 allocs/op,  14.4 tasks/op,   58212 insns/op,        0 errors)
median 905033.97 tps ( 72.3 allocs/op,  14.4 tasks/op,   58633 insns/op,        0 errors)

after:

median 945253.73 tps ( 72.2 allocs/op,  14.3 tasks/op,   57867 insns/op,        0 errors)
median 945324.68 tps ( 72.2 allocs/op,  14.3 tasks/op,   58086 insns/op,        0 errors)
median 945992.10 tps ( 72.2 allocs/op,  14.3 tasks/op,   58004 insns/op,        0 errors)
median 939955.27 tps ( 72.3 allocs/op,  14.3 tasks/op,   57971 insns/op,        0 errors)

No surprise

raphaelsc commented:
@bhalevy this will probably conflict with ur series to deglobalize repair history, so I'll rebase this series once yours get into master.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2165/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2167/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2190/

raphaelsc commented:
@scylladb/scylla-maint ping merge.
